### ISSUE TITLE
Building/Setting up lesson on Windows fails.

### ISSUE THREAD

**Issue by JoaoRodrigues** [carpentries/styles#445]
Hi there,

I'm trying to setup a new lesson on Windows 10 and I'm running into several issues that prevent me from generating the lesson website.

I am not running Git Bash nor WSL. I have both Ruby and Jekyll installed, and I installed make for Windows from the GnuWin32 page. I imported the styles repo (https://github.com/joaorodrigues/intro-git-github) and cloned it locally. Here's a list of steps and the issues that I found (with some workarounds):

1. Running `python bin/lesson_initialize.py` works and copies all the files:
```
_config.yml  _episodes_rmd/  _includes/  aio.md   AUTHORS  CITATION  CODE_OF_CONDUCT.md  data/  files/    LICENSE.md  README.md     setup.md
_episodes/   _extras/        _layouts/   assets/  bin/     code/     CONTRIBUTING.md     fig/   index.md  Makefile    reference.md
```

2. Running `make lesson-check` fails with a Python error. The scripts define `/usr/bin/env python3`, they should define `python` since not everyone has a python3 alias.

3. Running `make lesson-check` still fails, this time with an encoding problem:
```
index.md:6:FIXME: home page introduction
setup.md:4:FIXME
_episodes/01-introduction.md:6:- "Key question (FIXME)"
_episodes/01-introduction.md:8:- "First learning objective. (FIXME)"
_episodes/01-introduction.md:10:- "First key point. Brief Answer to questions. (FIXME)"
_episodes/01-introduction.md:12:FIXME
reference.md:7:FIXME
_extras/discuss.md:4:FIXME
_extras/guide.md:4:FIXME
Exception in thread Thread-5:
Traceback (most recent call last):
  File "C:\Users\joaor\Miniconda3\envs\scw\lib\threading.py", line 917, in _bootstrap_inner
    self.run()
  File "C:\Users\joaor\Miniconda3\envs\scw\lib\threading.py", line 865, in run
    self._target(*self._args, **self._kwargs)
  File "C:\Users\joaor\Miniconda3\envs\scw\lib\subprocess.py", line 1238, in _readerthread
    buffer.append(fh.read())
  File "C:\Users\joaor\Miniconda3\envs\scw\lib\encodings\cp1252.py", line 23, in decode
    return codecs.charmap_decode(input,self.errors,decoding_table)[0]
UnicodeDecodeError: 'charmap' codec can't decode byte 0x90 in position 8243: character maps to <undefined>

Traceback (most recent call last):
  File "C:\Users\joaor\Work\sc-teaching\intro-git-github\bin\lesson_check.py", line 565, in <module>
    main()
  File "C:\Users\joaor\Work\sc-teaching\intro-git-github\bin\lesson_check.py", line 121, in main
    docs = read_all_markdown(args.source_dir, args.parser)
  File "C:\Users\joaor\Work\sc-teaching\intro-git-github\bin\lesson_check.py", line 265, in read_all_markdown
    data = read_markdown(parser, filename)
  File "C:\Users\joaor\Work\sc-teaching\intro-git-github\bin\util.py", line 121, in read_markdown
    stdout_data, stderr_data = p.communicate(body)
  File "C:\Users\joaor\Miniconda3\envs\scw\lib\subprocess.py", line 939, in communicate
    stdout, stderr = self._communicate(input, endtime, timeout)
  File "C:\Users\joaor\Miniconda3\envs\scw\lib\subprocess.py", line 1288, in _communicate
    stdout = stdout[0]
IndexError: list index out of range
make: *** [lesson-check] Error 1
```

This is fixed by forcing the encoding of the `body` string in `p.communicate()` on `utils.py` line 121: `p.communicate(str(body.encode('utf-8')))`

4. Running `make lesson-check` still fails, this time with the following error:
```
index.md:6:FIXME: home page introduction
setup.md:4:FIXME
_episodes/01-introduction.md:6:- "Key question (FIXME)"
_episodes/01-introduction.md:8:- "First learning objective. (FIXME)"
_episodes/01-introduction.md:10:- "First key point. Brief Answer to questions. (FIXME)"
_episodes/01-introduction.md:12:FIXME
reference.md:7:FIXME
_extras/discuss.md:4:FIXME
_extras/guide.md:4:FIXME
Episode .\_episodes\01-introduction.md has badly-formatted filename
Missing required file ./CODE_OF_CONDUCT.md
Missing required file ./CONTRIBUTING.md
Missing required file ./LICENSE.md
Missing required file ./README.md
Missing required file ./_extras/discuss.md
Missing required file ./_extras/guide.md
Missing required file ./index.md
Missing required file ./reference.md
Missing required file ./setup.md
.\CONTRIBUTING.md: Internally-defined links may be missing definitions: "Data Carpentry"=>"dc-issues", "Data Carpentry"=>"dc-site", "GitHub flow"=>"github-flow", "GitHub"=>"github", "How to Contribute to an Open Source Project on GitHub"=>"how-contribute", "Library Carpentry"=>"lc-issues", "Library Carpentry"=>"lc-site", "Software Carpentry"=>"swc-issues", "Software Carpentry"=>"swc-site", "The Carpentries"=>"c-site", "bug reports"=>"issues", "create one"=>"github-join", "creating an issue"=>"issues", "discussion mailing list"=>"discuss-list"
.\LICENSE.md: Internally-defined links may be missing definitions: "Community Initiatives"=>"CI"
.\README.md: Internally-defined links may be missing definitions: "issues"=>"FIXME"
make: *** [lesson-check] Error 1
```

All the 'missing required files' are in this directory (see first item on this list).

I'm a bit stumped here. Don't really know how to progress. Would be great to have a few pointers of where to look. Happy to debug.

**Comment by fmichonneau** [comment]
I tested the script on your repo on my Linux system and I'm not seeing these errors, which makes me think that it's an operating system issue... A few things to check if you haven't already:
- is the script looking in the correct directory?
- is "/" directory separator recognized in your system?

@maxim-belkin or @brownsarahm do you have any other ideas?

**Comment by brownsarahm** [comment]
I also tried it quickly on linux and got different errors, then didn't have enough time to dig much deeper. I haven't tried running the script on any other repos recently, so I could have other things that are out of date. 

I also wonder why he needed to hard code in an encoding that was not needed before. If that's a system specific issue, we should figure out a way of using packages to try to avoid that. 

Most of the paths are done with `os.path.join` to be os agnostic, but the required files list has `/` in it. Could that be a part of the problem?

**Comment by maxim-belkin** [comment]
I can not troubleshoot this with the details provided.
João, could you please provide step-by-step instructions to reproduce this? That is:
- which gnuwin32 did you install?
- how did you install Ruby and Jekyll?
- ...

**Comment by JoaoRodrigues** [comment]
Hi @maxim-belkin 

### Installation and System Setup
I created a new user account on my windows machine to help debug this. I installed Ruby v2.6.5-1 using [RubyInstaller](https://rubyinstaller.org/downloads/) with the DevKit option. I then installed and used bundler (v 2.0.2) to install jekyll (version 3.8.5 to be compatible with Github pages).

I installed GNU Make from [here](http://gnuwin32.sourceforge.net/packages/make.htm), version is:
```
C:\Users\SoftwareCarpentry\Work\sc-teaching\intro-git-github $ "C:\Program Files (x86)\GnuWin32\bin\make.exe"  --version
GNU Make 3.81
Copyright (C) 2006  Free Software Foundation, Inc.
This is free software; see the source for copying conditions.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.

This program built for i386-pc-mingw32
```

### Setting up the carpentries lesson website

Re-running the setup instructions leads to the same problems. Now with logs.

```C:\Users\SoftwareCarpentry\Work\sc-teaching\intro-git-github $ "C:\Program Files (x86)\GnuWin32\bin\make.exe"  lesson-check
index.md:6:FIXME: home page introduction
index.md:14:> FIXME
setup.md:4:FIXME
_episodes/01-introduction.md:6:- "Key question (FIXME)"
_episodes/01-introduction.md:8:- "First learning objective. (FIXME)"
_episodes/01-introduction.md:10:- "First key point. Brief Answer to questions. (FIXME)"
_episodes/01-introduction.md:12:FIXME
reference.md:7:FIXME
_extras/discuss.md:4:FIXME
_extras/guide.md:4:FIXME
env: 'python3': No such file or directory
make: *** [lesson-check] Error 127```

Solved by fixing the shebang line on every Python script in `bin/` from `#!/usr/bin/env python3` to `#!/usr/bin/env python`.

Then I get the encoding problem. From this [StackOverflow answer](https://stackoverflow.com/a/33291200), this seems to be a by-product of using `universal_newlines=True`. This has also been reported as a bug in Python as well (report [here](https://bugs.python.org/issue27179)). We can get around it by specifying the encoding directly, which is what I did previously, by changing line 121 to:

```
body = str(body.encode('utf-8'))
p.communicate(body)
````

Now I get a little further:

```
C:\Users\SoftwareCarpentry\Work\sc-teaching\intro-git-github $ "C:\Program Files (x86)\GnuWin32\bin\make.exe"  lesson-check
index.md:6:FIXME: home page introduction
index.md:14:> FIXME
setup.md:4:FIXME
_episodes/01-introduction.md:6:- "Key question (FIXME)"
_episodes/01-introduction.md:8:- "First learning objective. (FIXME)"
_episodes/01-introduction.md:10:- "First key point. Brief Answer to questions. (FIXME)"
_episodes/01-introduction.md:12:FIXME
reference.md:7:FIXME
_extras/discuss.md:4:FIXME
_extras/guide.md:4:FIXME
Episode .\_episodes\01-introduction.md has badly-formatted filename
Missing required file ./CODE_OF_CONDUCT.md
Missing required file ./CONTRIBUTING.md
Missing required file ./LICENSE.md
Missing required file ./README.md
Missing required file ./_extras/discuss.md
Missing required file ./_extras/guide.md
Missing required file ./index.md
Missing required file ./reference.md
Missing required file ./setup.md
.\CONTRIBUTING.md: Internally-defined links may be missing definitions: "Data Carpentry"=>"dc-issues", "Data Carpentry"=>"dc-site", "GitHub flow"=>"github-flow", "GitHub"=>"github", "How to Contribute to an Open Source Project on GitHub"=>"how-contribute", "Library Carpentry"=>"lc-issues", "Library Carpentry"=>"lc-site", "Software Carpentry"=>"swc-issues", "Software Carpentry"=>"swc-site", "The Carpentries"=>"c-site", "bug reports"=>"issues", "create one"=>"github-join", "creating an issue"=>"issues", "discussion mailing list"=>"discuss-list"
.\LICENSE.md: Internally-defined links may be missing definitions: "Community Initiatives"=>"CI"
.\README.md: Internally-defined links may be missing definitions: "issues"=>"FIXME"
make: *** [lesson-check] Error 1
```

First problem: missing files when they exist. 

@fmichonneau, answering your question. `source_dir` in `bin/lesson_check.py` is set to `'.'`, as default. Indeed, having `/` as a separator is a bad idea for cross-platform compatibility ;)  The code is also very hard to follow to be honest. I refactored the `read_all_markdown` function and changed the `REQUIRED_FILES` variable and `check_fileset` function as well. Now `os.path.join` is used everywhere instead of hard-coding slashes in paths.

Second problem: `Episode .\_episodes\01-introduction.md has badly-formatted filename`. I modified the regular expression to remove the directory name and just split the path in the function before running regex.search.

Third problem: invalid links. I can probably fix these myself. I am just trying to get things up and running, without changing anything in the files. If this is not the case, this should be the case in my opinion. I think it's much nicer to start with an empty template (or some lorem ipsum text for placeholders) than having to add content from the beginning just to serve the site in jekyll.

I managed to get around this last problem by modying the default value of `permissive` in the `lesson_check.py` script. Why are you using Makefiles for this? I'm sorry, I know this sounds frustrated, but I've setup several Jekyll websites (from templates) and I never had all these issues. This setup sounds very convoluted..

Finally, `make serve` fails because of a comment (of all things). The error is:
```C:\Users\SoftwareCarpentry\Work\sc-teaching\intro-git-github $ "C:\Program Files (x86)\GnuWin32\bin\make.exe"  serve
jekyll serve
Configuration file: C:/Users/SoftwareCarpentry/Work/sc-teaching/intro-git-github/_config.yml
            Source: C:/Users/SoftwareCarpentry/Work/sc-teaching/intro-git-github
       Destination: C:/Users/SoftwareCarpentry/Work/sc-teaching/intro-git-github/_site
 Incremental build: disabled. Enable with --incremental
      Generating...
  Liquid Exception: Invalid syntax for include tag: teaching_time=e.teaching ... Valid syntax: {% include file.ext param='value' param2='value' %} in /_layouts/episode.html
jekyll 3.8.5 | Error:  Invalid syntax for include tag:

  teaching_time=e.teaching ...

Valid syntax:

  {% include file.ext param='value' param2='value' %}


make: *** [serve] Error 1
```

And the offending line is an include statement inside a comment on `_includes/episode_overview.html` (right in the header). If I remove that line, my local site finally renders. Weirdly enough, if I put it inside `{% raw %}` tags (inside the comment) it also works...

I collected all these changes in a bunch of commits and separate branch in my own fork of the 'styles' repo. I will make a pull request that you all can test on your own machines and see if it improves cross-compatibility.

**Comment by fmichonneau** [comment]
@JoaoRodrigues, has PR #446 addressed the issues you brought up here?

**Comment by maxim-belkin** [comment]
Closing this as there has been no actuvity in 3 months. Joao, feel free to create a new issue or comment here again! Thanks for the report!
